### PR TITLE
[feat] Add `{,load_}state_dict` to `ResultCollection` 1/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Tracking Issue: https://github.com/PyTorchLightning/pytorch-lightning/issues/7898

This PR adds a mechanism to reload from state_dict and restore logged values. 

Fixes #<issue_number>

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
> This PR adds a mechanism to reload from state_dict and restore logged values.

Could you describe more what is the problem we're trying to solve? Is this the only option for restoring logged values? 

Relying on `self.log` to checkpoint metric states is adding even more responsibilities when I think we should be going the other direction and moving responsibilities out of self.log.

**Comment by tchaton** [comment]
> > This PR adds a mechanism to reload from state_dict and restore logged values.
> 
> Could you describe more what is the problem we're trying to solve? Is this the only option for restoring logged values?
> 
> Relying on `self.log` to checkpoint metric states is adding even more responsibilities when I think we should be going the other direction and moving responsibilities out of self.log. [#7183 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/issues/7183#issuecomment-833182143)

Hey @ananthsub,

Currently, the ResultCollection isn't fault tolerant. 

Problems:
* On failure, the collection is being lost.
* On restoration, the ResultMetric lost reference to the Metrics.

In order to resolve this, this PR adds the following:

* A state_dict and load_from_state_dict function to the ResultCollection to restore in case of failure.
* A new field in the metadata object called `attribute_name`. This is used to automatically re-create references to the Metric on reload. 

The restoration isn't implemented yet, it would be the 2/n and 3/n PRs.

The `self.log` function of the `LightningModule` is responsible to properly populate the Metadata by extracting the attribute_name. I don't believe the `self.log` function is doing more than it should right now.

The state dumping / restoration will be added to `CheckpointConnector`. There will be a `restore_result_collections`.

PR 2/n: Add `ResultCollection` state_dict of all loops to model `checkpoint` inside CheckpointConnector  `dump_checkpoint` function

PR 3/n: Add `ResultCollection` restore within restoring calls.

Best,
T.C

**Comment by tchaton** [comment]
Hey Ananth, actually it might be possible to get rid of the attritube name. I will try this on Monday, which will reduce self.log responsability

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -505,3 +528,60 @@ def __getstate__(self) -> dict:
...
+                    apply_to_collection(item, ResultMetric, to_state_dict), metadata=item.meta
+                )
+            return _ResultMetricSerializationHelper(**item.__getstate__())
+
+        return {
+            k: apply_to_collection(v, (ResultMetric, ResultMetricCollection), to_state_dict)
+            for k, v in self.items()
+        }
+
+    def load_from_state_dict(self, state_dict: Dict[str, Any], metrics: Optional[Dict[str, Metric]] = None) -> None:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py]: Thanks ! Done !

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -505,3 +531,54 @@ def __getstate__(self) -> dict:
...
+        def to_state_dict(
+            item: Union[ResultMetric, ResultMetricCollection]
+        ) -> Union[_ResultMetricSerializationHelper, _ResultMetricCollectionSerializationHelper]:
+            if isinstance(item, ResultMetricCollection):
+                return _ResultMetricCollectionSerializationHelper(
+                    apply_to_collection(item, ResultMetric, to_state_dict), metadata=item.meta
+                )
+            state = item.__getstate__()
+            state["meta"].sync.fn = None
+            return _ResultMetricSerializationHelper(**item.__getstate__())
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:546]: is there no way to avoid pickling one of our objects?
If  one day the implementation or even just the name or import changes, it will not be possible to load the checkpoints, which has given us serious headaches already with callbacks, etc.

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:546]: Great question ! 
@awaelchli While applying `apply_to_collection`, we needed a placeholder to differentiate from pure dict.
I couldn't find a better idea. Do you have one ?

Right now, `_ResultMetricSerializationHelper` should an blank dictionarry and the name shouldn't change.

```py
class _ResultMetricSerializationHelper(dict):    
       """    Since ``ResultCollection`` can hold ``ResultMetric`` values or dictionaries of them, we need    a class to       differentiate between the cases after converting to state dict when saving its state.    """
```

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:546]: @carmocca Any thoughts ?

File: tests/core/test_metric_result_integration.py [hunk 3]
```diff
@@ -175,3 +184,109 @@ def lightning_log(fx, *args, **kwargs):
...
+            result = ResultCollection(True, torch.device("cpu"))
+            result.load_state_dict(state_dict, sync_fn=_result['training_step.a'].meta.sync.fn)
+
+            assert _result.items() == result.items()
+            assert _result["training_step.c_1"].meta == result["training_step.c_1"].meta
+
+        batch_idx = None
+
+        epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
+        _epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
```
Comment (awaelchli) [inline tests/core/test_metric_result_integration.py:246]:
```suggestion
        epoch_log = result.metrics(on_step=False)[MetricSource.LOG]
        _epoch_log = _result.metrics(on_step=False)[MetricSource.LOG]
```
Maybe this was the intention?
Anyway it's not so clear why we test that the outputs of a deepcloned result is the same. 
maybe put a comment what the difference is between _result and result ?

**Review by awaelchli — APPROVED** [review]
apart from the pickling concern LGTM
confidence low, not yet so familiar with new logger connector etc.

File: tests/core/test_metric_result_integration.py (hunk 3 above)
Comment (tchaton) [reply tests/core/test_metric_result_integration.py]: Good catch ! Renamed result_copy for clarity.

**Review by ethanwharris — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
Again low confidence but the plan makes sense and test for restoration looks good

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 4]
```diff
@@ -201,6 +202,9 @@ def __repr__(self) -> str:
             state += f", cumulated_batch_size={self.cumulated_batch_size}"
         return f"{self.__class__.__name__}({state})"
 
+    def __getstate__(self) -> dict:
+        return {**super().__getstate__(), '_class': self.__class__.__name__}
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:206]: Should we care about pickling the `_Metadata` and `_Sync` dataclasses?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 5]
```diff
@@ -501,7 +530,37 @@ def __str__(self) -> str:
...
+            elif cls == ResultMetricCollection.__name__:
+                result_metric = ResultMetricCollection()
+            else:
+                raise ValueError(f"Unexpected class name: {cls}")
+            result_metric.__setstate__(item)
+            return result_metric
+
+        items = {k: setstate(v) for k, v in state['items'].items()}
+        self.update(items)
+
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:561]:
```suggestion
        items = {k: setstate(v) for k, v in state['items'].items()}
        self.update(items)
        
        self.to(self.device)
```

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:206]: I think so. And we should drop the sync.fn function too.

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:206]: Done

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:561]: Done. Also implemented `map_location` and added a test for it

**Review by carmocca — APPROVED** [review]
